pkg/dispers/conceptindexor.go

```diff
@@ -0,0 +1,200 @@
+}
+
+// ID is used to get ID
+func (t *ConceptDoc) ID() string {
+	return t.ConceptID
+}
+
+// Rev is used to get Rev
+func (t *ConceptDoc) Rev() string {
+	return t.ConceptRev
+}
+
+// DocType is used to get DocType
+func (t *ConceptDoc) DocType() string {
+	return doctypeSalt
+}
+
+// Clone is used to create another ConceptDoc from this ConceptDoc
+func (t *ConceptDoc) Clone() couchdb.Doc {
+	cloned := *t
+	return &cloned
+}
+
+// SetID is used to set the Doc ID
+func (t *ConceptDoc) SetID(id string) {
+	t.ConceptID = id
+}
+
+// SetRev is used to set Rev
+func (t *ConceptDoc) SetRev(rev string) {
+	t.ConceptRev = rev
+}
+
+// addSalt does not check if one salt is already existing
+// It only creates a doc in ConceptIndexor's database
+func addSalt(concept string) error {
+
+	ConceptDoc := &ConceptDoc{
+		ConceptID:  "",
+		ConceptRev: "",
+		Concept:    concept,
+		Salt:       string(crypto.GenerateRandomBytes(defaultLenSalt)),
+	}
+	return couchdb.CreateDoc(prefixerCI, ConceptDoc)
+}
+
+func getSalt(concept string) (string, error) {
+
+	// Create index to run a mango query
+	salt := ""
+	err := couchdb.DefineIndex(prefixerCI, mango.IndexOnFields(doctypeSalt, "concept-index", []string{"concept"}))
+	if err != nil {
+		return salt, err
+	}
+
+	// Precise and run the mango query
+	var out []ConceptDoc
+	req := &couchdb.FindRequest{Selector: mango.Equal("concept", concept)}
+	err = couchdb.FindDocs(prefixerCI, doctypeSalt, req, &out)
+	if err != nil {
+		return salt, err
+	}
+
+	// Create a salt if no salt in the database for this concept
+	if len(out) == 1 {
+		salt = out[0].Salt
+	} else if len(out) == 0 {
+		return "", errors.New("Concept Indexor : no existing salt for this concept")
+	} else {
+		return "", errors.New("Concept Indexor : several salts for this concept")
+	}
+
+	return salt, err
+}
+
+func hash(str string, salt string) (string, error) {
+
+	// scrypt the cleartext string with the same parameters
+	other, err := scrypt.Key([]byte(str), []byte(salt), defaultN, defaultR, defaultP, defaultDkLen)
+	if err != nil {
+		return "", err
+	}
+	return string(other), err
+}
+
+// isConceptExisting is not useful to create a concept
+// isConceptExisting is useful to test CreateConcept or DeleteConcept
+func isConceptExisting(concept string) (bool, error) {
+
+	err := couchdb.DefineIndex(prefixerCI, mango.IndexOnFields(doctypeSalt, "concept-index", []string{"concept"}))
+	if err != nil {
+		return false, err
+	}
+
+	var out []ConceptDoc
+	req := &couchdb.FindRequest{Selector: mango.Equal("concept", concept)}
+	err = couchdb.FindDocs(prefixerCI, doctypeSalt, req, &out)
+	if err != nil {
+		return false, err
+	}
+
+	if len(out) > 0 {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+// CreateConcept is used to get a concept's salt. If the salt is absent from CI database
+// the function creates the salt and notify the user that the salt has just been created
+func CreateConcept(in *dispers.Concept) error {
```

Comment:
The `CreateConcept` goal should be to create the concept, not to get the salt. So for me, the function should be something like:
1. Check if concept exists in db
2. If yes, return an error saying the concept already exists and therefore cannot be added.
3. If no, create the salt, save the concept in db and return the hash.

Reply:
It is a very old comment. I update that. 

CreateConcept works like that : 
1. Check if concept exists in db
2. If yes, get the salt and return the hash
3. If no, create the salt, save the concept in db and return the hash.

That's why I did not like the name CreateConcept, I wanted to build a function that return a hash in everycase (except the one where there are too much salt in database).

Reply:
You are trying to do two different things here:
* create a new concept
* get the hash for an existing concept

Those occurs at different moments in dispers, so this should be splitted in 2 differents functions called by 2 different routes